docs/development/pigs/pig-029.rst

```diff
@@ -0,0 +1,207 @@
+********************************
+
+* Author: Atreyee Sinha, Maxime Regard
+* Created: ``2024-12-13``
+* Accepted: ``xx``
+* Status: xx
+* Discussion: `#4381`_
+
+Abstract
+========
+
+This PIG intends to resturcture `gammapy.modelling.models` to allow for flexible model definitions on arbitrary variables.
+
+Motivation
+==========
+
+Currently, gammapy supports Temporal, Spectral and Spatial models, which are bundled together in the `SkyModel` class for modelling and fitting. Despite the underlying `maps` framework being completely flexible, the modelling framework follows a fixed API allowing only spectral (compulsory), temporal and spatial models. If a user wants to model the behaviour in other parameters, say phase or frequency, it requires expert knowledge and is often performed using other fitting packages, eg: see https://github.com/cgalelli/gammapy_SyLC/tree/main/gammapy_SyLC.
+
+Currently, the `MapDataset` and the `FluxPointsDataset` support additional axes, so the main changes are required in the `MapEvaluator` and in the model definitions
+
+
+Use cases
+=========
+
+1. Fitting other parameters like phase, frequency, etc
+2. Reduction of code duplication: Analytic shapes like the power law is defined separately for `PowerLawSpectralModel`, `PowerLawTeLmporalModel` and `PowerLawNormSpectralModel`, which differ only in the unit handling
+3. Separation of units and values: This is a side benefit which can be useful if move to using JAX arrays in the future.
+
+
+Proposed API
+============
+
+We propose to have a general unit-less analytic class definition, inheriting from `ModelBase` The unit handling responsibility lies on the user.
+Internal support and checks and be provided for common input parameters (eg: time/energy) and the expected output units. 
+
+So, we move from 
+
+`SkyModel(spectral_model=, temporal_model=)`
+to 
+
+`SkyModelGeneral(submodels={"energy_true": p, "time":t})`
+
+where the user defines the axes the models are supposed affect.
+
+
+For a simple power law, it will be:
+
+"$\\phi(x) = \\phi_0 \\cdot \\left( \\frac{x-x_{ref}}{x_0} \\right)^{-\\alpha}$\n",
+
+
+```
+def integrate_linear(func, range_min, range_max, ndecade=100, oversampling_factor=100, **kwargs):
+        if isinstance(range_min, Time):
+            range_min=range_min.mjd * u.d
+            range_max=range_max.mjd * u.d
+        x, steps = np.linspace(
+            range_min, range_max, oversampling_factor, retstep=True, axis=-1
+        )
+        values = func(x)
+        integral = np.sum(values, axis=-1) * steps
+        return integral/np.sum(range_max - range_min) 
+
+class PowerLawModel(ModelBase):
+
+    tag = "Generalised_po"
+    alpha = Parameter("alpha", 2, frozen=True)
+    x0 = Parameter("x0", 1, frozen=True)
+    x_ref = Parameter("x_ref", 1, frozen=True)
+    phi0 = Parameter("phi0", 1, frozen=True, scale_method="scale10",
+                    interp="log")
+    
+    def __init__(self, input_units="", norm_units="", **kwargs):
```

Comment:
why do you need input_units="", norm_units="" ?
they can be determined from the unit of the parameters given in input

Reply:
Yes! But I thought its safer to set the units on init, rather than during the call so that things dont change silently.
Otherwise, you would suggest to set it during the first call to the function?